### DISCUSSION TITLE
How to get HTTP Headers? (newbie both in Rust & Rocket)

### DISCUSSION THREAD

**Discussion by DuredhelFinceleb** [rwf2/Rocket#2041] [Questions] [answered]
Sorry for the stupid question

I'm used to frameworks in other languages where I can just do something resembling this:
```
const authHeader = HTTP_HEADERS['Authorization'];
```

I could not find something built-in in Rocket 0.5.0-rc1 to do this

Do I really have to do this kind of boilerplate?
```
struct RequestHeaders<'h>(&'h HeaderMap<'h>);

#[rocket::async_trait]
impl<'r> FromRequest<'r> for RequestHeaders<'r> {
    type Error = Infallible;

    async fn from_request(request: &'r Request<'_>) -> request::Outcome<Self, Self::Error> {
        let request_headers = request.headers();
        Outcome::Success(RequestHeaders(request_headers))
    }
}

#[get("/")]
fn index(headers: RequestHeaders) -> String {
    format!("Hello, world ! {}", headers.0.get_one("authorization").unwrap_or(""))
}
```

As I wrote in the titile I'm a newbie both in Rust & Rocket so that code might not be very good, sorry about that

**Comment by IniterWorker** [comment]
Hi @DuredhelFinceleb,

In your use case, **Request Guards** are the answer to handle the validation process.

**For more information about request processing and guards:**
-  [how-rocket-works](https://rocket.rs/v0.5-rc/overview/#how-rocket-works)
-  [request-guards](https://rocket.rs/v0.5-rc/guide/requests/#request-guards)

You should **validate** HTTP Headers **in** the **request guard** as below:

```rust
#[rocket::async_trait]
impl<'r> FromRequest<'r> for Authenticated {
    type Error = ApiError;
    async fn from_request(req: &'r Request<'_>) -> request::Outcome<Self, Self::Error> {
        let res = header_bearer_validation(req).await;
        match res {
            Err(e) => request::Outcome::Failure((
                Status::from_code(e.status_code).unwrap_or(Status::InternalServerError),
                e,
            )),
            Ok(authenticated) => request::Outcome::Success(authenticated),
        }
    }
}
```

B. r

**Comment by DuredhelFinceleb** [comment, marked as answer]
Thanks a lot for your answer.
That's what I ended up doing for this case of Basic auth

But if I just want ot have a look at a header (for example I just want to log it) or perform a trivial process on it (for example if http_referer == "path" then do this else do that), do I have to write a request guard every time for every header (apart for the one supported out of the box)?

**Reply by IniterWorker** [reply]
Depending on your needs, If it is a global process you should consider [Fairings](https://rocket.rs/v0.5-rc/guide/fairings/#fairings).

> As a general rule of thumb, only globally applicable actions should be effected through fairings. You should not use a fairing to implement authentication or authorization (preferring to use a request guard instead) unless the authentication or authorization applies to all or the overwhelming majority application. On the other hand, you should use a fairing to record timing and usage statistics or to enforce global security policies.

**Reply by IniterWorker** [reply]
To answer the main questions in your thoughts :
- You need at least one Guard or a Fairing, in all your routes that require headers processing.
- One Request Guard can handle multiple headers if you want. ( You could return a struct with all the strings )
- Fairings can be set globally.

**Reply by DuredhelFinceleb** [reply]
> To answer the main questions in your thoughts :
> 
> * You need at least one Guard or a Fairing, in all your routes that require headers processing.
> * One Request Guard can handle multiple headers if you want. ( You could return a struct with all the strings )
> * Fairings can be set globally.

Thanks a lot this answers my question! 😃
